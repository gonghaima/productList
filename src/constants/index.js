import keyMirror from 'fbjs/lib/keyMirror';

/**
 * @namespace Constants
 * @desc App constants
 */

/**
 * @constant {Object} ActionTypes
 * @memberof Constants
 */
export const ActionTypes = keyMirror({
  SWITCH_MENU: undefined,
  EXCEPTION: undefined,
  PRODUCT_GET_REPOS: undefined,
  PRODUCT_GET_REPOS_SUCCESS: undefined,
  PRODUCT_GET_REPOS_FAILURE: undefined,
  SELECTION_CHANGE: undefined,
  PAGE_CHANGE: undefined,
});

/**
 * @constant {Object} STATUS
 * @memberof Constants
 */
export const STATUS = {
  IDLE: 'idle',
  RUNNING: 'running',
  READY: 'ready',
  SUCCESS: 'success',
  ERROR: 'error',
};
