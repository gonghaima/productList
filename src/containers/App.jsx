import React from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Router, Switch, Route } from 'react-router-dom';
import Helmet from 'react-helmet';
import styled, { ThemeProvider } from 'styled-components';

import history from 'modules/history';
import theme, { appColor, titleColor } from 'modules/theme';

import config from 'config';

import Home from 'routes/Home';
import NotFound from 'routes/NotFound';

import GlobalStyles from 'components/GlobalStyles';
import RouteProduct from 'components/RouteProduct';

const AppWrapper = styled.div`
  display: flex;
  background-color: ${appColor};
  flex-direction: column;
  min-height: 100vh;
  opacity: 1 !important;
  position: relative;
  transition: opacity 0.5s;
  h3,
  h5 {
    color: ${titleColor};
  }
`;

const Main = styled.main`
  min-height: 100vh;
`;

export class App extends React.Component {
  static propTypes = {};

  render() {
    return (
      <Router history={history}>
        <ThemeProvider theme={theme}>
          <AppWrapper>
            <Helmet
              defer={false}
              htmlAttributes={{ lang: 'pt-br' }}
              encodeSpecialCharacters={true}
              defaultTitle={config.title}
              titleTemplate={`%s | ${config.name}`}
              titleAttributes={{ itemprop: 'name', lang: 'pt-br' }}
            />
            <Main>
              <Switch>
                <RouteProduct path="/" component={Home} />
                <Route component={NotFound} />
              </Switch>
            </Main>
            <GlobalStyles />
          </AppWrapper>
        </ThemeProvider>
      </Router>
    );
  }
}

export default connect()(App);
