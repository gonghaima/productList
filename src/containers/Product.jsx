import React from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { appColor } from 'modules/theme';

import { getRepos } from 'actions/index';
import { STATUS } from 'constants/index';

import { Box, Flex, Heading, Link, Image, Paragraph, theme, utils } from 'styled-minimal';
import Loader from 'components/Loader';

const { responsive, spacer } = utils;
const { grays } = theme;

const ProductGrid = styled.ul`
  display: grid;
  grid-auto-flow: row;
  grid-gap: ${spacer(2)};
  grid-template-columns: 100%;
  list-style: none;
  margin: ${spacer(4)} auto 0;
  padding: 0;
  /* stylelint-disable */
  ${/* istanbul ignore next */ p =>
    responsive({
      ix: `
        grid-gap: ${spacer(3)(p)};
        width: 90%;
      `,
      md: `
        grid-template-columns: repeat(2, 1fr);
        width: 100%;
      `,
      lg: `
        grid-template-columns: repeat(3, 1fr);
      `,
      xl: `
        grid-gap: ${spacer(4)(p)};
        grid-template-columns: repeat(4, 1fr);
      `,
    })};
  /* stylelint-enable */

  > li {
    display: flex;
  }
`;

const Item = styled(Box)`
  align-items: center;
  border: solid 0.1rem ${appColor};
  border-radius: 0.4rem;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  overflow: hidden;
  padding: ${spacer(3)};
  text-align: center;
  width: 100%;
  /* stylelint-disable */
  ${/* istanbul ignore next */ p =>
    responsive({
      md: `
        padding: ${spacer(3)(p)};
      `,
      lg: `
        padding: ${spacer(4)(p)};
      `,
    })};
  /* stylelint-enable */

  > a {
    margin-bottom: ${spacer(2)};
  }

  img {
    height: 8rem;
  }
`;

const ItemHeader = styled.div`
  margin-bottom: ${spacer(3)};

  a {
    display: block;
  }

  small {
    color: ${grays.gray60};
  }
`;

const Pagination = styled.div`
  display: flex;
  flex-direction: row-reverse;
  ul {
    display: inline-block;
    padding-left: 15px;
    padding-right: 15px;
  }

  li {
    display: inline-block;
  }
  a {
    z-index: 3;
    color: #000;
    cursor: pointer;
    position: relative;
    float: left;
    padding: 6px 12px;
    margin-left: -1px;
    line-height: 1.42857143;
    text-decoration: none;
    background-color: #fff;
  }

  a:hover {
    border-bottom: 3px solid #337ab7;
  }
  a:focus {
    outline: unset;
    border-bottom: 6px solid #337ab7;
  }
`;

export class Product extends React.Component {
  state = {
    query: 'react',
  };

  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    product: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { query } = this.state;
    const { dispatch } = this.props;

    dispatch(getRepos(query));
  }

  render() {
    const { query } = this.state;
    const { product } = this.props;
    const data = product.repos.data[query] || [];
    let output;

    if (product.repos.status === STATUS.READY) {
      if (data.length) {
        output = (
          <ProductGrid data-type={query} data-testid="ProductGrid">
            {product.repos.data[query].map(d => (
              <li key={d.id}>
                <Item>
                  <Link href={d.html_url}>
                    <Image src={d.owner.avatar_url} alt={d.owner.login} />
                  </Link>
                  <ItemHeader>
                    <Link href={d.html_url}>
                      <Heading as="h5" lineHeight={1}>
                        {d.name}
                      </Heading>
                      <small>{d.owner.login}</small>
                    </Link>
                  </ItemHeader>
                  <Paragraph>{d.description}</Paragraph>
                </Item>
              </li>
            ))}
          </ProductGrid>
        );
      } else {
        output = <h3>Nothing found</h3>;
      }
    } else {
      output = <Loader block />;
    }

    return (
      <div key="Product" data-testid="GitHubWrapper">
        <Flex justifyContent="center" />
        {output}
        <Pagination>
          <ReactPaginate
            previousLabel={'< previous'}
            nextLabel={'next >'}
            breakLabel="..."
            breakClassName="break-me"
            pageCount={5}
            marginPagesDisplayed={2}
            pageRangeDisplayed={5}
            onPageChange={() => {}}
            containerClassName="pagination"
            subContainerClassName="pages pagination"
            activeClassName="active"
          />
        </Pagination>
      </div>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return { product: state.product };
}

export default connect(mapStateToProps)(Product);
