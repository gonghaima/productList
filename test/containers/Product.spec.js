import React from 'react';

import { STATUS } from 'constants/index';

import { Product } from 'containers/Product';

jest.mock('uuid/v4', () => () => 'ABCDE');

const mockDispatch = jest.fn();
const props = {
  dispatch: mockDispatch,
  product: {
    repos: {
      data: [],
    },
  },
};

function setup(ownProps = props) {
  return shallow(<Product {...ownProps} />, { attachTo: document.getElementById('react') });
}

describe('GitHub', () => {
  const wrapper = setup();

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Loader without data', () => {
    expect(wrapper.find('Loader')).toExist();
  });

  it('should have dispatched an action on mount', () => {
    expect(mockDispatch).toHaveBeenCalledWith({
      payload: { query: 'react' },
      type: 'PRODUCT_GET_REPOS',
    });
  });

  it("should not render if selected data doesn't exist", () => {
    wrapper.setProps({
      product: {
        repos: {
          data: {},
          status: STATUS.READY,
        },
      },
    });

    expect(wrapper.find('Grid')).not.toExist();
  });

  it('should render the Grid if data exists', () => {
    wrapper.setProps({
      product: {
        repos: {
          data: {
            react: [
              {
                id: 12,
                name: 'magic-tricks',
                html_url: 'https://github.com/username/magic-tricks',
                description: 'nothing much',
                owner: {
                  avatar_url: 'avatar_url',
                  login: 'username',
                },
              },
            ],
          },
          status: STATUS.READY,
        },
      },
    });

    expect(wrapper.find('Grid')).toMatchSnapshot();
  });
});
