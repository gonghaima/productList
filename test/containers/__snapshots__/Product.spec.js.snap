// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHub should render properly 1`] = `
<div
  data-testid="GitHubWrapper"
  key="Product"
>
  <Flex
    alignItems="center"
    justifyContent="center"
  />
  <Loader
    block={true}
    color="#00b4d5"
    size={32}
    type="grow"
  />
  <Pagination>
    <PaginationBoxView
      activeClassName="active"
      breakClassName="break-me"
      breakLabel="..."
      containerClassName="pagination"
      disableInitialCallback={false}
      disabledClassName="disabled"
      marginPagesDisplayed={2}
      nextClassName="next"
      nextLabel="next >"
      onPageChange={[Function]}
      pageCount={5}
      pageRangeDisplayed={5}
      previousClassName="previous"
      previousLabel=
< previous

      subContainerClassName="pages pagination"
    />
  </Pagination>
</div>
`;

exports[`GitHub should render the Grid if data exists 1`] = `null`;
